src/tools/rbd_mirror/ImageDeleter.cc

```diff
@@ -0,0 +1,536 @@
+    curr_deletion->on_delete = ctx;
+    m_delete_lock.Unlock();
+    return;
+  }
+
+  for (const auto& del_info : m_delete_queue) {
+    if (del_info->match(image_name)) {
+      del_info->on_delete = ctx;
+      m_delete_lock.Unlock();
+      return;
+    }
+  }
+
+  for (const auto& del_info : m_failed_queue) {
+    if (del_info->match(image_name)) {
+      del_info->on_delete = ctx;
+      m_delete_lock.Unlock();
+      return;
+    }
+  }
+
+  m_delete_lock.Unlock();
+
+  // image not scheduled for deletion
+  ctx->complete(0);
+}
+
+void ImageDeleter::process_image_delete() {
+
+  stringstream ss;
+  curr_deletion->to_string(ss);
+  std::string del_info_str = ss.str();
+  dout(10) << "start processing delete request: " << del_info_str << dendl;
+  int r;
+  cls::rbd::MirrorImage mirror_image;
+
+  // remote image was disabled, now we need to delete local image
+  IoCtx ioctx;
+  r = m_local->ioctx_create2(curr_deletion->local_pool_id, ioctx);
+  if (r < 0) {
+    derr << "error accessing local pool: " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(r);
+    return;
+  }
+
+  dout(20) << "connected to local pool: " << ioctx.get_pool_name() << dendl;
+
+  bool is_primary = false;
+  r = Journal<>::is_tag_owner(ioctx, curr_deletion->local_image_id, &is_primary);
+  if (r < 0 && r != -ENOENT) {
+    derr << "error retrieving image primary info: " << cpp_strerror(r)
+         << dendl;
+    enqueue_failed_delete(r);
+    return;
+  }
+  if (is_primary) {
+    dout(10) << "local image is the primary image, aborting deletion..."
+             << dendl;
+    m_delete_lock.Lock();
+    DeleteInfo *del_info = curr_deletion.release();
+    m_delete_lock.Unlock();
+    del_info->notify(-EISPRM);
+    return;
+  }
+
+  dout(20) << "local image is not the primary" << dendl;
+
+  bool has_snapshots;
+  r = image_has_snapshots_and_children(&ioctx, curr_deletion->local_image_id,
+                                       &has_snapshots);
+  if (r < 0) {
+    enqueue_failed_delete(r);
+    return;
+  }
+
+  mirror_image.global_image_id = curr_deletion->global_image_id;
+  mirror_image.state = cls::rbd::MIRROR_IMAGE_STATE_DISABLING;
+  r = cls_client::mirror_image_set(&ioctx, curr_deletion->local_image_id,
+                                           mirror_image);
+  if (r == -EEXIST || r == -EINVAL) {
+    derr << "cannot disable mirroring for image id" << curr_deletion->local_image_id
+         << ": global_image_id has changed/reused, aborting deletion: "
+         << cpp_strerror(r) << dendl;
+    m_delete_lock.Lock();
+    DeleteInfo *del_info = curr_deletion.release();
+    m_delete_lock.Unlock();
+    del_info->notify(r);
+    return;
+  } else if (r < 0) {
+    derr << "cannot disable mirroring for image id "
+         << curr_deletion->local_image_id << ": " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(r);
+    return ;
+  }
+
+  dout(20) << "setted local image mirroring to disable" << dendl;
```

Comment:
Minor: just saw this in the logs -- should be "set local image mirroring state to disabling"